src/core/agent/services/ipexCommunicationService.ts

```diff
@@ -103,6 +110,91 @@ class IpexCommunicationService extends AgentService {
     });
   }
 
+  async offerAcdc(notificationId: string) {
+    const keriNoti = await this.getNotificationRecordById(notificationId);
+    const msgSaid = keriNoti.a.d as string;
+    const msg = await this.signifyClient.exchanges().get(msgSaid);
+    const schemaSaid = msg.exn.a.s;
+    const creds = await this.signifyClient.credentials().list({
+      filter: {
+        "-s": { $eq: schemaSaid },
+      },
+    });
+    if (!creds || creds.length === 0) {
+      throw new Error(
+        IpexCommunicationService.CREDENTIAL_NOT_FOUND_WITH_SCHEMA
+      );
+    }
+    const pickedCred = creds[0];
+    let holderSignifyName;
+    const holder = await this.identifierStorage.getIdentifierMetadata(
+      msg.exn.a.i
+    );
+    if (holder && holder.signifyName) {
+      holderSignifyName = holder.signifyName;
+    } else {
+      const identifierHolder = await this.getIdentifierById(msg.exn.a.i);
+      holderSignifyName = identifierHolder?.name;
+    }
+    if (!holderSignifyName) {
+      throw new Error(IpexCommunicationService.AID_NOT_FOUND);
+    }
+
+    const [offer, sigs, gend] = await this.signifyClient.ipex().offer({
+      senderName: holderSignifyName,
+      recipient: msg.exn.i,
+      acdc: new Serder(pickedCred),
+    });
+    await this.signifyClient
+      .ipex()
+      .submitOffer(holderSignifyName, offer, sigs, gend, [msg.exn.i]);
+    await this.notificationStorage.deleteById(notificationId);
+  }
+
+  async grantApplyAcdc(notificationId: string) {
+    const keriNoti = await this.getNotificationRecordById(notificationId);
+    const msgSaid = keriNoti.a.d as string;
+    const msg = await this.signifyClient.exchanges().get(msgSaid);
+    const exnMessage = JSON.parse(msg.exn.a.m);
```

Comment:
We need to make sure the `offer` we are receiving is in reply to an `apply` message sent by us.

They could send us an `offer` message linked to an `apply` message signed by anyone. Please investigate this

Reply:
Sorry I used the wrong names there. We need to make sure the `agree` we are receiving is in response to an `offer` we sent.

IPEX messages are chained together, but maybe KERIA already does these checks for us.

Can you do some testing here and see what happens if a signify connection sends us an agree message that's linked to an offer that we DIDN'T send? If we don't receive the notification/message then I think we can assume this is already taken care of. But if not we will need to use the SAIDs within the messages to find the old message.

Reply:
Yep, I did some hack with signify-ts and send agree message to the another people who didn't send the offer message before, and it doenn't receive the agree message although the operation of send agree msg is done.

Reply:
Hmm, yeah I know they need to be linked correctly. But can do you do this exactly (since this is valid IPEX):
1. Send apply to WalletA
2. WalletA sends offer
3. Send offer (linked to offer from 2) but to WalletB
4. Does it show up for WalletB?

Reply:
that  mean when it received the offer message, it will send the agree message to walletB?

Reply:
Yes sorry, updated that. So the agree message is VALID (linked to offer sent to walletA) except it was accidentally sent to walletB. So this is exactly what you mean?

(And not that the agree message was invalid / not linked to a valid offer message)

Reply:
Yep. Exactly what I mean. 
The agree message need to be initiaiized with recepient and and offerSaid. So I guess that Keria checked the offer message is linked with offer of wallet

Reply:
https://github.com/WebOfTrust/keripy/blob/main/src/keri/vc/protocoling.py#L67-L68

OK, another test... because it seems it might just check that it exists in the wallet.

What if you:
1. Send apply to WalletA and WalletB
2. WalletA sends offer (WalletB does nothing)
3. Send agree (linked to offer from WalletA) but to walletB
4. Does it show up for WalletB?

If so this is more of a KERIA issue but I just want to be sure.

Reply:
It is nothing show up in walletB. Cool.